cext/cutils.c

```diff
@@ -0,0 +1,186 @@
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include <Python.h>
+
+#ifdef _MSC_VER
+
+typedef __int32 int32_t;
+typedef unsigned __int32 uint32_t;
+typedef __int64 int64_t;
+typedef unsigned __int64 uint64_t;
+
+#else
+#include <stdint.h>
+#endif
+
+static int32_t FNV1_OFFSET_BASIS = 0x811c9dc5;
+static int32_t FNV1_PRIME = 0x01000193;
+
+
+PyObject* hashcode(PyObject* self, PyObject *args);
+PyObject* schema_id(PyObject* self, PyObject *args);
+
+PyObject* str_hashcode(PyObject* data);
+int32_t str_hashcode_(PyObject* data, int lower);
+PyObject* b_hashcode(PyObject* data);
+
+static PyMethodDef methods[] = {
+    {"hashcode", (PyCFunction) hashcode, METH_VARARGS, ""},
+    {"schema_id", (PyCFunction) schema_id, METH_VARARGS, ""},
+    {NULL, NULL, 0, NULL}       /* Sentinel */
+};
+
+static struct PyModuleDef moduledef = {
+    PyModuleDef_HEAD_INIT,
+    "_cutils",
+    0,                    /* m_doc */
+    -1,                   /* m_size */
+    methods,              /* m_methods */
+    NULL,                 /* m_slots */
+    NULL,                 /* m_traverse */
+    NULL,                 /* m_clear */
+    NULL,                 /* m_free */
+};
+
+static char* hashcode_input_err = "supported only strings, bytearrays, bytes and memoryview";
+static char* schema_id_input_err = "input argument must be dict or int";
+static char* schema_field_type_err = "schema keys must be strings";
+
+PyMODINIT_FUNC PyInit__cutils(void) {
+	return PyModule_Create(&moduledef);
+}
+
+PyObject* hashcode(PyObject* self, PyObject *args) {
+    PyObject* data;
+
+    if (!PyArg_ParseTuple(args, "O", &data)) {
+        return NULL;
+    }
+
+    if (data == Py_None) {
+        return PyLong_FromLong(0);
+    }
+    else if (PyUnicode_CheckExact(data)) {
+        return str_hashcode(data);
+    }
+    else {
+        return b_hashcode(data);
+    }
+}
+
+PyObject* str_hashcode(PyObject* data) {
+    return PyLong_FromLong(str_hashcode_(data, 0));
+}
+
+int32_t str_hashcode_(PyObject *data, int lower) {
+    int32_t res = 0;
+    Py_ssize_t sz = PyUnicode_GET_LENGTH(data);
+
+    Py_ssize_t i;
+    for (i = 0; i < sz; i++) {
+        Py_UCS4 ch = PyUnicode_READ_CHAR(data, i);
+
+        if (lower) {
+            ch = Py_UNICODE_TOLOWER(ch);
+        }
```

Comment:
I'm not sure how fast `PyUnicode_READ_CHAR` and `Py_UNICODE_TOLOWER` are, but I think it may be faster to get the full string in one call using `PyUnicode_DATA` and work with C representation after that. Or at least we may want to use `PyUnicode_READ` instead of `PyUnicode_READ_CHAR`, as documentation states it is more efficient.

Reply:
Good point about PyUnicode_READ!